fast64_internal/utility.py

```diff
@@ -523,6 +525,118 @@ def deleteIfFound(filePath, stringValue):
 			fileData.write(stringData)
 		fileData.close()
 
+def yield_children(obj: bpy.types.Object):
+	yield obj
+	if obj.children:
+		for o in obj.children:
+			yield from yield_children(o)
+
+def iter_from_active():
+	active_obj = bpy.context.view_layer.objects.active # keep original obj reference
+	yield from yield_children(active_obj)
+	bpy.context.view_layer.objects.active = active_obj
+
+def store_original_mtx():
+	for obj in iter_from_active():
+		obj['original_mtx'] = obj.matrix_local
+
+def rotate_bounds(bounds, mtx: mathutils.Matrix):
+	return [
+		(mathutils.Vector(b[:]) @ mtx).to_tuple()
+		for b in bounds[:]
+	]
+
+def obj_scale_is_unified(obj):
+	'''Combine scale values into a set to ensure all values are the same'''
+	return len(set(obj.scale)) == 1
+
+def translation_rotation_from_mtx(mtx: mathutils.Matrix):
+	'''Strip scale from matrix'''
+	t, r, _ = mtx.decompose()
+	return Matrix.Translation(t).to_4x4() @ r.to_matrix().to_4x4()
+
+def scale_mtx_from_vector(scale: mathutils.Vector):
+	scax = mathutils.Matrix.Scale(scale[0], 4, (1, 0, 0))
+	scay = mathutils.Matrix.Scale(scale[1], 4, (0, 1, 0))
+	scaz = mathutils.Matrix.Scale(scale[2], 4, (0, 0, 1))
+	return scax @ scay @ scaz
+
+def copy_object_and_apply(obj: bpy.types.Object, apply_scale = False, apply_modifiers = False):
+	if apply_scale or apply_modifiers:
+		# its a unique mesh, use object name
+		obj['instanced_mesh_name'] = obj.name
+		obj.original_name = obj.name
+		if apply_scale:
+			obj['original_mtx'] = translation_rotation_from_mtx(mathutils.Matrix(obj['original_mtx']))
+
+	obj_copy = obj.copy()
+	obj_copy.parent = None
+	# reset transformations
+	obj_copy.location = mathutils.Vector([0.0, 0.0, 0.0])
+	obj_copy.scale = mathutils.Vector([1.0, 1.0, 1.0])
+	obj_copy.rotation_quaternion = mathutils.Quaternion([1, 0, 0, 0])
+	obj_copy.data = obj_copy.data.copy()
+
+	if apply_modifiers:
+		# In order to correctly apply modifiers, we have to go through blender and add the object to the collection, then apply modifiers
+		prev_active = bpy.context.view_layer.objects.active
+		bpy.context.collection.objects.link(obj_copy)
+		obj_copy.select_set(True)
+		bpy.context.view_layer.objects.active = obj_copy
+		for modifier in obj_copy.modifiers:
+			attemptModifierApply(modifier)
+
+		bpy.context.view_layer.objects.active = prev_active
+
+	mtx = mathutils.Quaternion((1, 0, 0), math.radians(-90.0)).to_matrix().to_4x4()
+	if apply_scale:
+	 	mtx = mtx @ scale_mtx_from_vector(obj.scale)
+
+	obj_copy.data.transform(mtx)
+	# Flag used for finding these temp objects
+	obj_copy['temp_export'] = True
+
+	# Override for F3D culling bounds (used in addCullCommand)
+	bounds_mtx = scale_mtx_from_vector([1, -1, -1]) # reverse y/z
+	if apply_scale:
+		bounds_mtx *= scale_mtx_from_vector(obj.scale) # apply scale if needed
+	bounds_mtx = bounds_mtx @ mathutils.Quaternion((1, 0, 0), math.radians(-90.0)).to_matrix().to_4x4()
```

Comment:
may want to store the previous matrix I commented on and reuse it here

Reply:
i was originally going to store it somewhere but the shared references are a big scary since the first variable `mtx` needs to be modified. i could make a lambda that returns a new one instead?

![image](https://user-images.githubusercontent.com/79979276/132070642-12657cea-aadb-4073-bdbb-9e07cfb606f7.png)